internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,250 @@
+//			f.buffer.WriteEntry(f.parser.Entry())
+//		}
+//		if err := f.parser.Err(); err != nil {
+//			logger.Noticef(
+//				"Forwarding logs to target %q: cannot parse log for service %q: %v",
+//				f.targetName, f.serviceName, err)
+//			// TODO: create a new parser?
+//		}
+//		if !f.buffer.ReadyToFlush() {
+//			f.flushBuffer()
+//		}
+//	}
+//
+//	// Forwarder stopped or service exited - flush remaining logs
+//	if len(f.buffer.messages) > 0 {
+//		f.flushBuffer()
+//	}
+//}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot send logs for service %q: %v",
+			f.targetName, f.serviceName, err)
+		// TODO: retry?
+	}
+
+	f.buffer.Reset()
+}
+
+type lokiBuffer struct {
+	labels map[string]string
+
+	buffer  *bytes.Buffer
+	encoder *json.Encoder
+
+	messages    int
+	maxMessages int
+	deadline    *time.Time
+	delay       time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	buf := bytes.NewBuffer(make([]byte, 0, 4096))
+
+	b := &lokiBuffer{
+		labels:      labels,
+		buffer:      buf,
+		encoder:     json.NewEncoder(buf),
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	b.Reset()
+	return b
+}
+
+func (b *lokiBuffer) Reset() {
+	b.messages = 0
+	b.deadline = nil
+	b.buffer.Reset()
+
+	// Write start of request
+	b.buffer.WriteString(`{"streams":[{"stream":`)
+	b.encoder.Encode(b.labels)
+	b.buffer.WriteString(`,"values":[`)
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    ["<unix epoch in nanoseconds>","<log line>"]
+	b.buffer.WriteString("[")
+	timestamp := strconv.FormatInt(entry.Time.UnixNano(), 10)
+	b.encoder.Encode(timestamp)
+	b.buffer.WriteString(",")
+	message := strings.Trim(entry.Message, "\n")
+	b.encoder.Encode(message)
+	b.buffer.WriteString("],")
+
+	b.messages++
+	if b.deadline == nil {
+		// Set write deadline
+		deadline := time.Now().Add(b.delay)
+		b.deadline = &deadline
+	}
+}
+
+func (b *lokiBuffer) ReadyToFlush() bool {
+	return (b.messages >= b.maxMessages) ||
+		(b.deadline != nil && time.Now().After(*b.deadline))
+	// TODO: flush when buffer length is past a certain threshold?
+}
+
+// Prepares a JSON request for sending to Loki.
+func (b *lokiBuffer) Request() io.Reader {
+	// Need to remove last ',' in buffer
+	b.buffer.Truncate(b.buffer.Len() - 1)
+	// Write end of request
+	b.buffer.WriteString("]}]}")
+
+	return b.buffer
+}
+
+const (
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient handles sending requests to Loki.
+type lokiClient struct {
+	client *http.Client
+	url    string
+}
+
+func newLokiClient(location string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{
+			Timeout: 10 * time.Second,
+		},
+		url: location,
+	}
+}
+
+func (c *lokiClient) Send(body io.Reader) error {
```

Comment:
I am not overly keen on the separation of concerns here. The loki client is the one that understands the loki protocol but here it's just accepting a bag of bytes that we assume matches "application/json" and fire it off.

Sure you can argue that these guys are in the same package and private so it's safe. But it's really not a well thought out design as we have no real guarantees around anything at the moment.

I would suggest that we don't try and decode our buffers into the Loki format but that instead the client takes vector of strongly typed log message and does it's own heavy lifting.

Reply:
Fixed, now we use the following interface for clients:
```go
type logClient interface {
	Send([]servicelog.Entry) error
}
```
and each client contains the backend-specific encoding logic, whereas the forwarder is generic.